### PR TITLE
[NVPTX][AMDGPU] Move annotation creation out of clang

### PR DESCRIPTION
**Pull request by frasercrmck**
This patch refactors the way we lower SYCL attributes and properties to NVVM annotations, through function attributes and metadata. It unifies the flow better with the SPIR-V paths at the same time.

Previously we had:

1. Clang handling function attributes in two places:
    1. CodeGenFunction did generic lowering of attributes to function metadata
    2. NVPTXTargetCodeGenInfo did its own additional lowering of attributes to NVVM annotations
2. Kernel properties being handled in clang. NVPTXTargetCodeGenInfo lowered kernel properties, which had already been converted to function attributes, to NVVM annotations.
3. Kernel properties for HIP/CUDA *not* being turned into function metadata. Because function metadata is how the ComputeModuleRuntimeInfo library creates runtime info, this meant that target backends couldn't act on properties, as they were lost during lowering.

Fundamentally, clang is not the correct place for handling SYCL kernel properties as it unnecessarily touches the front-end for what is really library code.

With this patch, we now have:

1. Clang handles function attributes in one place, lowering to function metadata
2. SYCLLowerIR lowers kernel properties to the same function metadata: now working for HIP and CUDA kernels too. Because function metadata is consistently present for kernel properties, they behave as expected on HIP/CUDA targets now.
3. A new pass converts function metadata to NVVM annotations, replacing the code we've taken out of clang.

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/TargetHelpers.cpp [hunk 1]
```diff
@@ -46,15 +47,74 @@ std::string getAnnotationString(ArchType AT) {
   return std::string();
 }
 
-void populateKernels(Module &M, SmallVectorImpl<KernelPayload> &Kernels,
-                     ArchType AT) {
+SmallPtrSet<Function *, 4> getKernels(Module &M) {
```
Comment (frasercrmck) [inline llvm/lib/SYCLLowerIR/TargetHelpers.cpp:50]: I'm not sure about this: using `amdgcn.annotations` to identify kernels seems unusual when it's our downstream addition, and they seem to have `amdgcn_kernel` calling conventions. It's how other passes in SYCLLowerIR do it so I blindly copied it. I think I'll try and make it use the calling convention instead, though. Unless there's a good reason for `amdgcn.annotations` I doubt it will make it upstream.

File: sycl/test-e2e/Basic/work_group_size_prop.cpp [hunk 2]
```diff
@@ -1,8 +1,5 @@
-// RUN: %{build} -o %t.out
+// RUN: %{build} --gcc-toolchain=/opt/slurm/gcc/13.2.0 -o %t.out
```
Comment (EwanC) [inline sycl/test-e2e/Basic/work_group_size_prop.cpp:1]: `--gcc-toolchain=/opt/slurm/gcc/13.2.0` is this addition intended to be committed, or an artifact from testing? If so could you elaborate on why we need it.

Comment (frasercrmck) [reply sycl/test-e2e/Basic/work_group_size_prop.cpp:1]: Whoops no that was local! Thanks for spotting that.

Comment (frasercrmck) [reply sycl/test-e2e/Basic/work_group_size_prop.cpp]: Pushed an undo for that

**Review by EwanC — APPROVED** [review]
Graph test changes LGTM

**Review by dm-vodopyanov — APPROVED** [review]
`sycl/test-e2e/*` LGTM

**Review by sarnex — APPROVED** [review]
llvm/lib/SYCLLowerIR/CMakeLists.txt lgtm

**Comment by bader** [comment]
@intel/dpcpp-cfe-reviewers, ping.

File: llvm/lib/SYCLLowerIR/TargetHelpers.cpp (hunk 1 above)
Comment (frasercrmck) [reply llvm/lib/SYCLLowerIR/TargetHelpers.cpp:50]: See #14713 for a better method.

**Review by elizabethandrews — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by bader** [comment]
@frasercrmck, please, resolve merge conflicts.

File: llvm/lib/SYCLLowerIR/TargetHelpers.cpp (hunk 1 above)
Comment (frasercrmck) [reply llvm/lib/SYCLLowerIR/TargetHelpers.cpp]: This has gone

**Comment by frasercrmck** [comment]
> @frasercrmck, please, resolve merge conflicts.

Done, thanks. I haven't waited for all CI but Linux looks happy at least.

**Comment by frasercrmck** [comment]
@intel/llvm-gatekeepers this looks ready to merge, thank you.

**Comment by sarnex** [comment]
@frasercrmck We are seeing macos post commit failures, can you please take a look?

```
FAILED: lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o 
ccache /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/runner/work/llvm/llvm/build/lib/SYCLLowerIR -I/Users/runner/work/llvm/llvm/src/llvm/lib/SYCLLowerIR -I/Users/runner/work/llvm/llvm/build/include -I/Users/runner/work/llvm/llvm/src/llvm/include -I/Users/runner/work/llvm/llvm/src/llvm/projects/vc-intrinsics/GenXIntrinsics/include -I/Users/runner/work/llvm/llvm/build/projects/vc-intrinsics/GenXIntrinsics/include -I/Users/runner/work/llvm/llvm/build/_deps/vc-intrinsics-src/GenXIntrinsics/include -I/Users/runner/work/llvm/llvm/build/_deps/vc-intrinsics-build/GenXIntrinsics/lib/GenXIntrinsics/../../include -I/Users/runner/work/llvm/llvm/build/_deps/vc-intrinsics-build/GenXIntrinsics/include -isystem /usr/local/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -std=c++17 -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.7  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o -MF lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o.d -o lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o -c /Users/runner/work/llvm/llvm/src/llvm/lib/SYCLLowerIR/SYCLCreateNVVMAnnotations.cpp
/Users/runner/work/llvm/llvm/src/llvm/lib/SYCLLowerIR/SYCLCreateNVVMAnnotations.cpp:62:10: error: no viable conversion from returned value of type 'array<optional<size_t>, [...]>' to function return type 'array<optional<uint64_t>, [...]>'
  return Ops;
         ^~~
/Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/array:129:29: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::array<std::optional<size_t>, 3>' (aka 'array<optional<unsigned long>, 3>') to 'const std::array<std::optional<unsigned long long>, 3> &' for 1st argument
struct _LIBCPP_TEMPLATE_VIS array
                            ^
/Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/array:129:29: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'std::array<std::optional<size_t>, 3>' (aka 'array<optional<unsigned long>, 3>') to 'std::array<std::optional<unsigned long long>, 3> &&' for 1st argument
1 error generated.
```

**Comment by frasercrmck** [comment]
> @frasercrmck We are seeing macos post commit failures, can you please take a look?
> 
> https://github.com/intel/llvm/actions/runs/10253109905/job/28365021433
> 
> ```
> FAILED: lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o 
> ccache /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/runner/work/llvm/llvm/build/lib/SYCLLowerIR -I/Users/runner/work/llvm/llvm/src/llvm/lib/SYCLLowerIR -I/Users/runner/work/llvm/llvm/build/include -I/Users/runner/work/llvm/llvm/src/llvm/include -I/Users/runner/work/llvm/llvm/src/llvm/projects/vc-intrinsics/GenXIntrinsics/include -I/Users/runner/work/llvm/llvm/build/projects/vc-intrinsics/GenXIntrinsics/include -I/Users/runner/work/llvm/llvm/build/_deps/vc-intrinsics-src/GenXIntrinsics/include -I/Users/runner/work/llvm/llvm/build/_deps/vc-intrinsics-build/GenXIntrinsics/lib/GenXIntrinsics/../../include -I/Users/runner/work/llvm/llvm/build/_deps/vc-intrinsics-build/GenXIntrinsics/include -isystem /usr/local/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -std=c++17 -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.7  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o -MF lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o.d -o lib/SYCLLowerIR/CMakeFiles/LLVMSYCLLowerIR.dir/SYCLCreateNVVMAnnotations.cpp.o -c /Users/runner/work/llvm/llvm/src/llvm/lib/SYCLLowerIR/SYCLCreateNVVMAnnotations.cpp
> /Users/runner/work/llvm/llvm/src/llvm/lib/SYCLLowerIR/SYCLCreateNVVMAnnotations.cpp:62:10: error: no viable conversion from returned value of type 'array<optional<size_t>, [...]>' to function return type 'array<optional<uint64_t>, [...]>'
>   return Ops;
>          ^~~
> /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/array:129:29: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::array<std::optional<size_t>, 3>' (aka 'array<optional<unsigned long>, 3>') to 'const std::array<std::optional<unsigned long long>, 3> &' for 1st argument
> struct _LIBCPP_TEMPLATE_VIS array
>                             ^
> /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/array:129:29: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'std::array<std::optional<size_t>, 3>' (aka 'array<optional<unsigned long>, 3>') to 'std::array<std::optional<unsigned long long>, 3> &&' for 1st argument
> 1 error generated.
> ```

Whoops, sorry. See #14964

**Comment by sarnex** [comment]
Thanks!
